src/braket/default_simulator/simulator.py

```diff
@@ -202,32 +212,96 @@ def _validate_and_consolidate_observable_result_types(
-                existing_obs = qubit_observable_mapping.get(qubit)
-                if existing_obs:
-                    if existing_obs != new_qubit_obs:
-                        raise ValueError(
-                            f"Qubit {qubit} is already being measured in {existing_obs};"
-                            f" cannot measure in {new_obs}."
-                        )
-                else:
-                    qubit_observable_mapping[qubit] = new_qubit_obs
-                    qubit_obs_obj = obs_obj[i] if isinstance(obs_obj, dict) else obs_obj
-                    if (
-                        not none_observable_mapping.get(new_qubit_obs)
-                        and qubit_obs_obj.measured_qubits.index(qubit) == 0
-                    ):
-                        qubit_obs_obj = obs_obj[i] if isinstance(obs_obj, dict) else obs_obj
-                        not_none_observable_list.append(qubit_obs_obj)
-        return not_none_observable_list + unique_none_observables
+                DefaultSimulator._assign_observable(
+                    obs_obj,
+                    new_obs,
+                    measured_qubits,
+                    i,
+                    not_none_observable_list,
+                    qubit_observable_mapping,
+                    none_observable_mapping,
+                    identity_qubits,
+                )
+        for i in sorted(identity_qubits):
+            not_none_observable_list.append(Identity([i]))
+        return not_none_observable_list
+
+    @staticmethod
+    def _assign_observable(
+        whole_observable,
+        hashed_qubit_observable,
+        measured_qubits,
+        target_index,
+        not_none_observable_list,
+        qubit_observable_mapping,
+        none_observable_mapping,
+        identity_qubits,
+    ):
+        # Validate that the same observable is requested for a qubit in the result types
+        hashed_qubit_observable = (
+            hashed_qubit_observable[target_index]
+            if isinstance(hashed_qubit_observable, dict)
+            else hashed_qubit_observable
+        )
+        qubit = measured_qubits[target_index]
+        existing_observable = qubit_observable_mapping.get(qubit)
+        if hashed_qubit_observable == Identity.__name__:
+            if qubit not in qubit_observable_mapping:
+                identity_qubits.add(qubit)
+            # Do nothing if non-identity observable already exists on the qubit
+        else:
+            qubit_observable = (
+                whole_observable[target_index]
+                if isinstance(whole_observable, dict)
+                else whole_observable
+            )
+            if not existing_observable:
+                identity_qubits.discard(qubit)
+                qubit_observable_mapping[qubit] = qubit_observable
+
+                if (
+                    not none_observable_mapping.get(hashed_qubit_observable)
+                    and qubit_observable.measured_qubits.index(qubit) == 0
```

Comment:
I spent a while being confused on this check, so a comment would probably help here.